Invoice print spec: added tax

filipefurtad0 opened the pull request:
#### What? Why?

Relates to #7983.

Aim is to capture the current behavior for on what concerns the displaying of taxes on both `legacy` and `alternative` invoices.

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

- changes the order and product factories to consider added (not-included) tax rates/amounts
- fixes broken specs due to factory change
- adds 4 test cases around these 4 invoices/test cases - displaying below for clarity:

  - included tax, legacy invoice
  - included tax, alternative invoice
  - added tax, legacy invoice
  - added tax, alternative invoice

![image](https://user-images.githubusercontent.com/49817236/183123171-31163b8f-c10b-47ff-8e9f-83db4a8711c6.png)

Yellow boxes refer to issue #7983.

Pending examples have been numbered i) to v) and referenced in the pic above and spec.

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

Green build.

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->

Relates to #7983.

#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

filipefurtad0 commented on spec/features/admin/invoice_print_spec.rb:
```diff
@@ -120,6 +123,237 @@
+
+      before do
+        order1.reload
+        break unless order1.next! until order1.delivery?
```
I'm not entirely sure how to address this one. Any suggestions?

jibees replied:
Could this be remplaced by 
```ruby
        until order1.delivery?
          break unless order1.next!
        end
```

But I think I do prefer the version:

```ruby
while !order1.delivery?
  break if !order1.next!
end
```

filipefurtad0 replied:
> But I think I do prefer the version:

Thanks JB. Commited and squashed this second suggestion:+1:

jibees reviewed (approved):
Thanks a lot for this one @filipefurtad0 ! 
It's very valuable to have this kind of spec, on which we can rely on in case of changing something. Thanks! 🙏 
I reviewed your PR, and didn't find anything wrong, but i must admit that it's difficult to read/understand all those values (GST, taxes, price, ...). Anyway, that's a (wonderful) starting point!

What's your point of view on how this can interact with #7983 ?

filipefurtad0 commented:
Thanks for the kind review :raised_hands: 

Yes, It's very difficult to read/follow the values... I've tried to simplify but it was quite challenging :sweat_smile:  so I ended up taking the data setup from [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/999b04ef9937cee900f614db44230c5c558b1e97/spec/system/admin/reports_spec.rb#L483).

The set up is repeated for two distinct orders (order1, order2) which corresponds to the two different tax scenarios (respectively: included tax, added tax). Many lines of code, again, maybe possible to simplify - but this way we are sure we have independent setups.

> What's your point of view on how this can interact with https://github.com/openfoodfoundation/openfoodnetwork/issues/7983 ?

There are pending examples should correspond to the yellow blocks on the pictures. Specifically, the `pending "closing #7983"` can be changed to pass when the issue is addressed.

The other pending examples relate to the fact that $0.0 tax are currently not displayed for line items or enterprise fees, but I was not sure this was under scope of #7983.

I totally agree this is a starting point - Surely it can be improved.

filipefurtad0 commented:
Hey @jibees ,

I've updated the picture above and added some other pending examples on the spec: they should match now. Hope this makes easier to read :ok_man:

mkllnk commented on spec/features/admin/invoice_print_spec.rb:
```diff
@@ -239,6 +239,119 @@
       end
+    end
+
+        context "added" do
```
Whitespace! :wink:

filipefurtad0 replied:
Thanks, incredible eye for detail :sparkles: 

This was corrected in the following commit :+1:

mkllnk reviewed (approved):
Excellent!

filipefurtad0 commented:
This spec is introducing flakiness, on the calculation of the total order tax - the values below are incorrect:

![image](https://user-images.githubusercontent.com/49817236/184121608-9b643c33-d577-4ffe-ac42-955520003404.png)

But this is not a real bug (from staging-FR) - correct tax total values:

![image](https://user-images.githubusercontent.com/49817236/184121759-f09e8dee-f5c5-4c19-bcb5-4db78af305bc.png)

I think this is caused by not explicitly loading `app/helpers/checkout_helper.rb`

```
  def display_checkout_tax_total(order)
    Spree::Money.new order.total_tax, currency: order.currency
  end
```

filipefurtad0 commented:
I'm blocked on this seemingly flaky spec. No idea why the invoice is not rendered correctly on the test environment, but correcltly in staging.

Opened this issue https://github.com/openfoodfoundation/openfoodnetwork/issues/9556.
